File: mockspresso-mockito/src/main/java/com/episode6/hackit/mockspresso/mockito/MockitoMockerConfig.java
package com.episode6.hackit.mockspresso.mockito;

import com.episode6.hackit.mockspresso.api.MockerConfig;
import org.mockito.Mock;
import org.mockito.Spy;

import java.lang.annotation.Annotation;
import java.util.Arrays;
import java.util.List;

/**
 * A MockerConfig for Mockito.
 *
 * @deprecated This functionality is now exposed by the kotlin extension method `mockByMockito()` and its
 * JavaSupport counterpart {@link MockspressoMockitoPluginsJavaSupport#mockByMockito()}
 *
 * This class will be marked internal/protected in a future release
 */
@Deprecated
public class MockitoMockerConfig implements MockerConfig {

  private final MockitoMockMaker mMockMaker = new MockitoMockMaker();
  private final MockitoFieldPreparer mFieldPreparer = new MockitoFieldPreparer();

  @Override
  public MockMaker provideMockMaker() {
    return mMockMaker;
  }

  @Override
  public FieldPreparer provideFieldPreparer() {
    return mFieldPreparer;
  }

  @Override
  public List<Class<? extends Annotation>> provideMockAnnotations() {
    return Arrays.asList(
        Mock.class,
        Spy.class);
  }

}

File: mockspresso-mockito/src/main/java/com/episode6/hackit/mockspresso/mockito/MockitoMockerConfig.kt
package com.episode6.hackit.mockspresso.mockito

import com.episode6.hackit.mockspresso.api.MockerConfig
import com.episode6.hackit.mockspresso.api.MockerConfig.FieldPreparer
import com.episode6.hackit.mockspresso.reflect.TypeToken
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.MockitoAnnotations
import org.mockito.Spy

/**
 * A MockerConfig for Mockito.
 *
 */
internal class MockitoMockerConfig : MockerConfig, MockerConfig.MockMaker, FieldPreparer {
  override fun provideMockMaker(): MockerConfig.MockMaker = this
  override fun provideFieldPreparer(): FieldPreparer = this
  override fun provideMockAnnotations(): List<Class<out Annotation?>> = listOf(Mock::class.java, Spy::class.java)


  override fun prepareFields(objectWithMockFields: Any) {
    @Suppress("DEPRECATION") // TODO: Define new mocker config that leverages openMocks properly
    MockitoAnnotations.initMocks(objectWithMockFields)
  }

  @Suppress("UNCHECKED_CAST") override fun <T> makeMock(typeToken: TypeToken<T>): T =
      Mockito.mock(typeToken.rawType as Class<T>)
}
